cirq-core/cirq/transformers/routing/line_initial_mapper.py

```diff
@@ -0,0 +1,237 @@
+
+"""Concrete implementation of AbstractInitialMapper that places lines of qubits onto the device."""
+
+from typing import Dict, Optional, TYPE_CHECKING
+import networkx as nx
+
+from cirq import value, _compat
+from cirq.transformers import routing
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class LineInitialMapper(routing.AbstractInitialMapper):
+    """Places logical qubits in the circuit onto physical qubits on the device."""
+
+    def __init__(self, device_graph: nx.Graph) -> None:
+        """Initializes a LineInitialMapper.
+
+        Args:
+            device_graph: device graph
+        """
+        # TODO: Tanuj, should this logic be done instead at the beginning of the routing transformer
+        # so that we don't have to repeat it for each InitialMapper and for the MappingManager?
+        if nx.is_directed(device_graph):
+            self.device_graph = nx.DiGraph()
+            self.device_graph.add_nodes_from(sorted(list(device_graph.nodes(data=True))))
+            self.device_graph.add_edges_from(sorted(list(device_graph.edges)))
+        else:
+            self.device_graph = nx.Graph()
+            self.device_graph.add_nodes_from(sorted(list(device_graph.nodes(data=True))))
+            self.device_graph.add_edges_from(
+                sorted(list(sorted(edge) for edge in device_graph.edges))
+            )
+
+    def _make_circuit_graph(self, circuit: 'cirq.AbstractCircuit') -> nx.Graph:
+        """Creates a (potentially incomplete) qubit connectivity graph of the circuit.
+
+        Iterates over the moments circuit from left to right drawing edges between logical qubits
+        that:
+            (1) have degree < 2, and
+            (2) that are involved in a 2-qubit operation in the current moment.
+        At this point the graph is forest of paths and/or simple cycles. For each simple cycle, make
+        it a path by removing the last edge that was added to it.
+
+        Args:
+            circuit: the input circuit with logical qubits
+
+        Returns:
+            The (potentially incomplete) qubit connectivity graph of the circuit.
+        """
+        circuit_graph = nx.Graph()
+        edge_order = 0
+        for op in circuit.all_operations():
+            circuit_graph.add_nodes_from(op.qubits)
+            if len(op.qubits) == 2 and all(
+                circuit_graph.degree[op.qubits[i]] < 2 for i in range(2)
+            ):
+                circuit_graph.add_edge(*op.qubits, edge_order=edge_order)
+                edge_order += 1
+        found = True
+        while found:
+            try:
+                cycle = nx.find_cycle(circuit_graph)
+                edge_to_remove = max(
+                    cycle, key=lambda x: circuit_graph.edges[x[0], x[1]]['edge_order']
+                )
+                circuit_graph.remove_edge(*edge_to_remove)
+            except nx.exception.NetworkXNoCycle:
+                found = False
+        return circuit_graph
```

Comment:
Let's use union find to make this cleaner and more efficient. 
```suggestion
    circuit_graph = nx.Graph()
    circuit_graph.add_nodes_from(circuit.all_qubits())
    union_find = nx.utils.UnionFind(elements=circuit.all_qubits())
    for op in circuit.all_operations():
        if (
            len(op.qubits) == 2
            and all(d < 2 for _, d in circuit_graph.degree(op.qubits))
            and union_find[op.qubits[0]] != union_find[op.qubits[1]]
        ):
            union_find.union(*op.qubits)
            circuit_graph.add_edge(*op.qubits)
    return circuit_graph
```